### PR TITLE
fix(1723): Change the query to the last updated

### PR DESCRIPTION
**Pull request by ibu1224**
## Context
This pr is that discussed in this issue.
https://github.com/screwdriver-cd/screwdriver/issues/1723

Current templates and commands page shows  the template with `MAX`.
It should shows template and command that sorted by createTime. However, in order to get the version number of the latest template, there are cases where it is not possible to use `createTime`, so I adopted `id`. I fixed to use below SQL.
```
> SELECT id, name, namespace ...
FROM templates 
WHERE templates.id = (
  SELECT MAX(t.id) 
  FROM templates AS t 
  WHERE templates.name = t.name 
  AND templates.namespace = t.namespace 
  AND templates.id <= t.id
) 
ORDER BY createTime ... ;

+----+-------+-------------+---------+------------+
| id | name  | namespace   | version | createTime |
+----+-------+-------------+---------+------------+
|  2 | foo   | bar         | 0.0.1   | NULL       |
+----+-------+-------------+---------+------------+
``` 

I use subquery only when there is arguments of groupBy.
And I put groupBy arguments in subquery `WHERE`.

## References
https://github.com/screwdriver-cd/screwdriver/issues/1723

## License

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

**Comment by yoshwata** [comment]
How about performance of this query? How much time spent for getting the result of query in your environment.

**Comment by ibu1224** [comment]
@tk3fftk 
OK. I change PR title.
I think that there is no problem because I don't use `groupBy` when use other table.

**Comment by ibu1224** [comment]
@yoshwata 
The performance is as follows:
**Current**
```
mysql> select count(id) from templates;
+-----------+
| count(id) |
+-----------+
|      6128 |
+-----------+
1 row in set (0.02 sec)

mysql> select max(name),max(namespace),max(version),max(createtime) from templates group by name, namespace order by max(createtime);
...
698 rows in set (0.08 sec)
```
**After change**
```
mysql> select count(id) from templates where templates.id = ( select max(t.id) from templates as t where templates.name = t.name and templates.namespace = t.namespace and templates.id <= t.id) order by createtime;
+-----------+
| count(id) |
+-----------+
|       698 |
+-----------+
1 row in set (0.20 sec)

```

**Comment by yoshwata** [comment]
Thanks @ibu1224. `getAllTemplates` spends 3.47ms in our environment. 
Screen Shot 2019-11-11 at 21 09 36

The effect is not so big as the total time to show list screen is over 7 sec.
Screen Shot 2019-11-11 at 21 18 11

Paginating is needed to get faster to show screen, but it is another context.

File: index.js [hunk 1]
```diff
@@ -472,10 +472,23 @@ class Squeakquel extends Datastore {
...
+            const where = { id: { [Sequelize.Op.gte]: Sequelize.col(`${config.table}.id`) } };
+
+            config.groupBy.forEach((v) => {
+                where[v] = { [Sequelize.Op.eq]: Sequelize.col(`${config.table}.${v}`) };
+            });
+            const subQuery = this.client.dialect.QueryGenerator.selectQuery(config.table, {
+                tableAs: 't',
+                attributes: [Sequelize.fn('MAX', Sequelize.col('t.id'))],
+                where
+            }).slice(0, -1);
```
Comment (tk3fftk) [inline index.js]: It removes `;` at the end of sting.  We should add a comment about it.

Comment (ibu1224) [reply index.js]: Thank you, I added  a comment.

**Comment by wahapo** [comment]
Would you give me a merge of this PR?
And then please merge https://github.com/screwdriver-cd/screwdriver/pull/1849, thanks.

**Comment by parthasl** [comment]
@yoshwata @wahapo @tk3fftk - templates and commands throw 500 internal server error in our beta environment. our beta environment tables are prefixed with beta-, please check and update if this works for templates and command tables prefixed with beta-

**Comment by wahapo** [comment]
@parthasl Thanks for feedback, I shared your feedback with the Screwdriver team in Japan.

**Comment by parthasl** [comment]
Thanks, @wahapo, we use Postgres database
errors: "invalid reference to FROM-clause entry for table \"templates\" and "invalid reference to FROM-clause entry for table \"commands\"".

**Comment by wahapo** [comment]
@parthasl I think the problem is because the subquery references `templates.name`. I want to isolate the cause. I expect that the same error will occur even in environments where without prefixed with `beta-`. Do you have the results you tried?

References
https://lets.postgresql.jp/documents/technical/lateral/1 (Japanese site)

**Comment by jithine** [comment]
@wahapo @parthasl this potentially fixed it https://github.com/screwdriver-cd/datastore-sequelize/pull/62/files

**Pull request merged**